### PR TITLE
Make Application also use DI

### PR DESCRIPTION
**Pull request by jorisvergeer**
Hi,

I was thinking that it would be nice if I could also define my own Application type. It would even be nicer if this Application could be created by the DI container. 

This patch does enables both.

What do you think?

### CODE REVIEW

**Comment by Lakritzator** [comment]
Hey Joris,

 you again 👍 

That looks useful, although I'm not 100% sure of the use cases.
The main thing that bothers me, is this syntax ".ConfigureWpf<MainWindow, MyApplication>()"
It feels so unnatural... it might be nicer to have a more fluent api.

I always need a bit of time to think about it, before I merge. Today I feel dead tired, due to my kid keeping me awake last night.

Thanks,
Robin

**Comment by jorisvergeer** [comment]
Hi,
Thanks for your reply.

My use case is that I would like to define themes in my own application xaml. It also feels like te appropriate place to put application wide components like a tray icon.

I can try to spend some time to create a more fluent API. 

Maybe something like this
```
ConfigureWpf(options => {
    options.UseMainWindow<MyMainWindow>();
    options.UseApplication<MyApplication>();
})
```

**Comment by jorisvergeer** [comment]
Hi, 

I changed the ConfigureWpf function to a more Fluent API. Which breaks the current API. Maybe you like it. :)

Joris

**Comment by Lakritzator** [comment]
Damn, I was merging the small changes from Simon Cropp and forgot that it cause a conflict.

I like the current API a lot better, but I am not so sure about the name of the UseMainWindow...

**Comment by jorisvergeer** [comment]
Do you like UseShell Better?

**Comment by Lakritzator** [comment]
I think it's better, and it should be possible & allowed to have multiple of them.
UseShellWindow would also be okay, maybe a bit much as it's all in a context where it should be clear.

File: src/Dapplo.Microsoft.Extensions.Hosting.Wpf/HostBuilderWpfExtensions.cs [hunk 1]
```diff
@@ -73,35 +77,51 @@ public static IHostBuilder ConfigureWpf(this IHostBuilder hostBuilder, Action<IW
...
-            hostBuilder.ConfigureWpf(configureAction);
-            hostBuilder.ConfigureServices((hostBuilderContext, serviceCollection) =>
+            if (wpfBuilder.ApplicationType != null)
             {
-                serviceCollection.AddSingleton<TView>();
+                // Check if the registered application does inherit System.Windows.Application
+                var baseApplicationType = typeof(Application);
+                if (!baseApplicationType.IsAssignableFrom(wpfBuilder.ApplicationType))
+                {
+                    throw new ArgumentException("The registered Application type inherit System.Windows.Application", nameof(configureDelegate));
```
Comment (Lakritzator) [inline src/Dapplo.Microsoft.Extensions.Hosting.Wpf/HostBuilderWpfExtensions.cs]: Shouldn't that be: The registered Application type **must** inherit System....

File: src/Dapplo.Microsoft.Extensions.Hosting.Wpf/HostBuilderWpfExtensions.cs [hunk 2]
```diff
@@ -73,35 +77,51 @@ public static IHostBuilder ConfigureWpf(this IHostBuilder hostBuilder, Action<IW
...
+                    serviceCollection.AddSingleton(wpfBuilder.ApplicationType);
+
+                    if (wpfBuilder.ApplicationType != baseApplicationType)
+                    {
+                        serviceCollection.AddSingleton(serviceProvider => (Application)serviceProvider.GetRequiredService(wpfBuilder.ApplicationType));
+                    }
+                });
+            }
+
+            if (wpfBuilder.MainWindowType != null)
```
Comment (Lakritzator) [inline src/Dapplo.Microsoft.Extensions.Hosting.Wpf/HostBuilderWpfExtensions.cs]: Make it possible to supply multiple, from my point of view a shell doesn't need to be limited to one window.

File: src/Dapplo.Microsoft.Extensions.Hosting.Wpf/IWpfBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Dapplo.Microsoft.Extensions.Hosting.Wpf
+{
+    /// <summary>
+    /// Interface used for configuring Wpf 
+    /// </summary>
+    public interface IWpfBuilder
```
Comment (Lakritzator) [inline src/Dapplo.Microsoft.Extensions.Hosting.Wpf/IWpfBuilder.cs]: The name is not 100% clear, we don't build Wpf but a WPF Configuration.
IWpfConfigurationBuilder might be more suitable.

**Pull request merged**